ocaml/xapi-cli-server/record_util.ml

```diff
@@ -1075,89 +1075,35 @@ let domain_type_of_string x =
 let vtpm_operation_to_string (op : API.vtpm_operations) =
   match op with `destroy -> "destroy"
 
+(** parse [0-9]*(b|bytes|kib|mib|gib|tib)* to bytes *)
+let bytes_of_string str =
+  let ( ** ) a b = Int64.mul a b in
+  let invalid msg = raise (Invalid_argument msg) in
+  try
+    Scanf.sscanf str "%Ld%s" @@ fun size suffix ->
+    match String.lowercase_ascii suffix with
+    | "bytes" | "b" | "" ->
+        size
+    | "kib" | "kb" | "k" ->
+        size ** 1024L
+    | "mib" | "mb" | "m" ->
+        size ** 1024L ** 1024L
+    | "gib" | "gb" | "g" ->
+        size ** 1024L ** 1024L ** 1024L
+    | "tib" | "tb" | "t" ->
+        size ** 1024L ** 1024L ** 1024L ** 1024L
+    | _ ->
+        invalid suffix
+  with _ -> invalid str
+
 (** Parse a string which might have a units suffix on the end *)
 let bytes_of_string field x =
-  let ( ** ) a b = Int64.mul a b in
-  let max_size_TiB =
-    Int64.div Int64.max_int (1024L ** 1024L ** 1024L ** 1024L)
-  in
-  (* detect big number that cannot be represented by Int64. *)
-  let int64_of_string s =
-    try Int64.of_string s
-    with _ ->
-      if s = "" then
-        record_failure
-          "Failed to parse field '%s': expecting an integer (possibly with \
-           suffix)"
-          field ;
-
-      let alldigit = ref true and i = ref (String.length s - 1) in
-      while !alldigit && !i > 0 do
-        alldigit := Astring.Char.Ascii.is_digit s.[!i] ;
-        decr i
-      done ;
-      if !alldigit then
-        record_failure
-          "Failed to parse field '%s': number too big (maximum = %Ld TiB)" field
-          max_size_TiB
```

Comment:
Is it not important anymore to distinguish between "not an integer - couldn't parse" and "integer - but too big"? Additionally, we could overflow above when multiplying ** 1024 several times, do we want to catch and report that?

Reply:
Overflow is detected and results in an exception, which is caught. That number is incredibly large, though.